In [4]:
# 📚 Basic Libraries
import pandas as pd # manipulación de datos
import numpy as np # matemáticas operaciones
import warnings # warnings son inútiles
import os

# 📊 Visualizations
import matplotlib.pyplot as plt # visualización básica
import seaborn as sns # visualización más chachi

# 🤖 Machine Learning
from sklearn.model_selection import train_test_split
from sklearn.metrics import roc_curve, confusion_matrix, ConfusionMatrixDisplay
from sklearn.metrics import classification_report

# ⚙️ Settings
pd.set_option('display.max_columns', None) # display all columns
warnings.filterwarnings('ignore') # ignore warnings

In [5]:
# 🔧 Basic functions
def snake_columns(data):
    """
    Standarize and returns snake_case columns
    """
    data.columns = [column.lower().replace(' ', '_') for column in data.columns]


# Challenge 1 - Import and Describe the Dataset

#### In this challenge we will use the `austin_weather` data.

#### First, import it into a data frame called `austin`.

In [6]:
# Import data:
os.listdir()
data = pd.read_csv('austin_weather.csv')
data.head()

Date  TempHighF  TempAvgF  TempLowF DewPointHighF DewPointAvgF  \
0  2013-12-21         74        60        45            67           49   
1  2013-12-22         56        48        39            43           36   
2  2013-12-23         58        45        32            31           27   
3  2013-12-24         61        46        31            36           28   
4  2013-12-25         58        50        41            44           40   

  DewPointLowF HumidityHighPercent HumidityAvgPercent HumidityLowPercent  \
0           43                  93                 75                 57   
1           28                  93                 68                 43   
2           23                  76                 52                 27   
3           21                  89                 56                 22   
4           36                  86                 71                 56   

  SeaLevelPressureHighInches SeaLevelPressureAvgInches  \
0                      29.86                     29.68   
1                      30.41                     30.13   
2                      30.56                     30.49   
3                      30.56                     30.45   
4                      30.41                     30.33   

  SeaLevelPressureLowInches VisibilityHighMiles VisibilityAvgMiles  \
0                     29.59                  10                  7   
1                     29.87                  10                 10   
2                     30.41                  10                 10   
3                      30.3                  10                 10   
4                     30.27                  10                 10   

  VisibilityLowMiles WindHighMPH WindAvgMPH WindGustMPH  \
0                  2          20          4          31   
1                  5          16          6          25   
2                 10           8          3          12   
3                  7          12          4          20   
4                  7          10          2          16   

  PrecipitationSumInches               Events  
0                   0.46  Rain , Thunderstorm  
1                      0                       
2                      0                       
3                      0                       
4                      T

#### Next, describe the dataset you have loaded:
- Look at the variables and their types
- Examine the descriptive statistics of the numeric variables
- Look at the first five rows of all variables to evaluate the categorical variables as well

## Explore data:

In [7]:
data.shape

(1319, 21)

In [8]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1319 entries, 0 to 1318
Data columns (total 21 columns):
 #   Column                      Non-Null Count  Dtype 
---  ------                      --------------  ----- 
 0   Date                        1319 non-null   object
 1   TempHighF                   1319 non-null   int64 
 2   TempAvgF                    1319 non-null   int64 
 3   TempLowF                    1319 non-null   int64 
 4   DewPointHighF               1319 non-null   object
 5   DewPointAvgF                1319 non-null   object
 6   DewPointLowF                1319 non-null   object
 7   HumidityHighPercent         1319 non-null   object
 8   HumidityAvgPercent          1319 non-null   object
 9   HumidityLowPercent          1319 non-null   object
 10  SeaLevelPressureHighInches  1319 non-null   object
 11  SeaLevelPressureAvgInches   1319 non-null   object
 12  SeaLevelPressureLowInches   1319 non-null   object
 13  VisibilityHighMiles         1319 non-null   obje

In [9]:
# Apply the function to change the column names
snake_columns(data)

In [10]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1319 entries, 0 to 1318
Data columns (total 21 columns):
 #   Column                      Non-Null Count  Dtype 
---  ------                      --------------  ----- 
 0   date                        1319 non-null   object
 1   temphighf                   1319 non-null   int64 
 2   tempavgf                    1319 non-null   int64 
 3   templowf                    1319 non-null   int64 
 4   dewpointhighf               1319 non-null   object
 5   dewpointavgf                1319 non-null   object
 6   dewpointlowf                1319 non-null   object
 7   humidityhighpercent         1319 non-null   object
 8   humidityavgpercent          1319 non-null   object
 9   humiditylowpercent          1319 non-null   object
 10  sealevelpressurehighinches  1319 non-null   object
 11  sealevelpressureavginches   1319 non-null   object
 12  sealevelpressurelowinches   1319 non-null   object
 13  visibilityhighmiles         1319 non-null   obje

In [11]:
# Create data copy:
df = data.copy()

In [12]:
df.describe().T

count       mean        std   min   25%   50%   75%    max
temphighf  1319.0  80.862775  14.766523  32.0  72.0  83.0  92.0  107.0
tempavgf   1319.0  70.642911  14.045904  29.0  62.0  73.0  83.0   93.0
templowf   1319.0  59.902957  14.190648  19.0  49.0  63.0  73.0   81.0

#### Given the information you have learned from examining the dataset, write down three insights about the data in a markdown cell below

#### Your Insights:

1. There are 21 variables in the dataset. 3 of them are numeric and the rest contain some text.

2. The average temperature in Austin ranged between around 70 degrees F and around 93 degrees F. The highest temperature observed during this period was 107 degrees F and the lowest was 19 degrees F.

3. When we look at the head function, we see that a lot of variables contain numeric data even though these columns are of object type. This means we might have to do some data cleansing.


#### Let's examine the DewPointAvgF variable by using the `unique()` function to list all unique values in this dataframe.

Describe what you find in a markdown cell below the code. What did you notice? What do you think made Pandas to treat this column as *object* instead of *int64*?

In [13]:
# Check column 'dewpointavgf'""
df.dewpointavgf.nunique()

66

In [14]:
df.dewpointavgf.isnull().sum()

0

In [15]:
df.dewpointavgf.unique() # dewpointavgf is being treated as obeject because of the '_'

array(['49', '36', '27', '28', '40', '39', '41', '26', '42', '22', '48',
       '32', '8', '11', '45', '55', '61', '37', '47', '25', '23', '20',
       '33', '30', '29', '17', '14', '13', '54', '59', '15', '24', '34',
       '35', '57', '50', '53', '60', '46', '56', '51', '31', '38', '62',
       '43', '63', '64', '67', '66', '58', '70', '68', '65', '69', '71',
       '72', '-', '73', '74', '21', '44', '52', '12', '75', '76', '18'],
      dtype=object)


---


* **Conclusions:** `dewpointavgf` is being treated as obeject because of the '_'


---


The following is a list of columns misrepresented as `object`. Use this list to convert the columns to numeric using the `pandas.to_numeric` function in the next cell. If you encounter errors in converting strings to numeric values, you need to catch those errors and force the conversion by supplying `errors='coerce'` as an argument for `pandas.to_numeric`. Coercing will replace non-convertable elements with `NaN` which represents an undefined numeric value. This makes it possible for us to conveniently handle missing values in subsequent data processing.

*Hint: you may use a loop to change one column at a time but it is more efficient to use `apply`.*

In [16]:
wrong_type_columns = ['dewpointhighf', 'dewpointavgf', 'dewpointlowf', 'humidityhighpercent',
                     'humidityavgpercent', 'humiditylowpercent', 'sealevelpressurehighinches',
                     'sealevelpressureavginches', 'sealevelpressurelowinches', 'visibilityhighmiles',
                     'visibilityavgmiles', 'visibilitylowmiles', 'windhighmph', 'windavgmph',
                     'windgustmph', 'precipitationsuminches']

for column in wrong_type_columns:
    df[column] = pd.to_numeric(df[column], errors='coerce')

#### Check if your code has worked by printing the data types again. You should see only two `object` columns (`Date` and `Events`) now. All other columns should be `int64` or `float64`.

In [17]:
# Your code here
df.dtypes

date                           object
temphighf                       int64
tempavgf                        int64
templowf                        int64
dewpointhighf                 float64
dewpointavgf                  float64
dewpointlowf                  float64
humidityhighpercent           float64
humidityavgpercent            float64
humiditylowpercent            float64
sealevelpressurehighinches    float64
sealevelpressureavginches     float64
sealevelpressurelowinches     float64
visibilityhighmiles           float64
visibilityavgmiles            float64
visibilitylowmiles            float64
windhighmph                   float64
windavgmph                    float64
windgustmph                   float64
precipitationsuminches        float64
events                         object
dtype: object

# Challenge 2 - Handle the Missing Data

#### Now that we have fixed the type mismatch, let's address the missing data.

By coercing the columns to numeric, we have created `NaN` for each cell containing characters. We should choose a strategy to address these missing data.

The first step is to examine how many rows contain missing data.

We check how much missing data we have by applying the `.isnull()` function to our dataset. To find the rows with missing data in any of its cells, we apply `.any(axis=1)` to the function. `austin.isnull().any(axis=1)` will return a column containing true if the row contains at least one missing value and false otherwise. Therefore we must subset our dataframe with this column. This will give us all rows with at least one missing value.

#### In the next cell, identify all rows containing at least one missing value. Assign the dataframes with missing values to a variable called `missing_values`.

In [18]:
# rows containing at least one missing value:
missing_values = df[df.isnull().any(axis=1)]
missing_values.sample(4)

date  temphighf  tempavgf  templowf  dewpointhighf  dewpointavgf  \
1125  2017-01-19         71        62        52           55.0          52.0   
842   2016-04-11         89        80        70           71.0          68.0   
1313  2017-07-26        103        91        78           75.0          70.0   
280   2014-09-27         80        74        68           69.0          66.0   

      dewpointlowf  humidityhighpercent  humidityavgpercent  \
1125          50.0                100.0                79.0   
842           66.0                 93.0                74.0   
1313          61.0                 91.0                58.0   
280           64.0                 93.0                78.0   

      humiditylowpercent  sealevelpressurehighinches  \
1125                57.0                       29.99   
842                 55.0                       29.98   
1313                25.0                       30.13   
280                 62.0                       30.05   

      sealevelpressureavginches  sealevelpressurelowinches  \
1125                      29.83                      29.70   
842                       29.89                      29.83   
1313                      30.03                      29.93   
280                       30.00                      29.95   

      visibilityhighmiles  visibilityavgmiles  visibilitylowmiles  \
1125                 10.0                10.0                10.0   
842                  10.0                10.0                 6.0   
1313                 10.0                10.0                10.0   
280                  10.0                10.0                 8.0   

      windhighmph  windavgmph  windgustmph  precipitationsuminches events  
1125         12.0         3.0         18.0                     NaN   Rain  
842          15.0         5.0         26.0                     NaN         
1313         15.0         6.0         27.0                     NaN         
280           9.0         2.0         14.0                     NaN   Rain

There are multiple strategies to handle missing data. Below lists the most common ones data scientists use:

* Removing all rows or all columns containing missing data. This is the simplest strategy. It may work in some cases but not others.

* Filling all missing values with a placeholder value.
    * For categorical data, `0`, `-1`, and `9999` are some commonly used placeholder values.
    * For continuous data, some may opt to fill all missing data with the mean. This strategy is not optimal since it can increase the fit of the model.

* Filling the values using some algorithm.

#### In our case, we will use a hybrid approach which is to first remove the data that contain most missing values then fill in the rest of the missing values with the *linear interpolation* algorithm.

#### Next, count the number of rows of `austin` and `missing_values`.

In [19]:
# Your code here
len(missing_values)

136

#### Calculate the ratio of missing rows to total rows

In [20]:
df.shape

(1319, 21)

In [21]:
# Ratio of missing rows to total rows:
pct_missing_rows = (len(missing_values)*100)/df.shape[0]
pct_missing_rows

10.310841546626232

As you can see, there is a large proportion of missing data (over 10%). Perhaps we should evaluate which columns have the most missing data and remove those columns. For the remaining columns, we will perform a linear approximation of the missing data.

We can find the number of missing rows in each column using the `.isna()` function. We then chain the `.sum` function to the `.isna()` function and find the number of missing rows per column

In [22]:
# Your code here
num_missing_rows = df.isna().sum().sort_values(ascending=False)
num_missing_rows

precipitationsuminches        124
visibilityhighmiles            12
visibilityavgmiles             12
visibilitylowmiles             12
dewpointhighf                   7
dewpointavgf                    7
dewpointlowf                    7
windgustmph                     4
sealevelpressurehighinches      3
sealevelpressureavginches       3
sealevelpressurelowinches       3
windhighmph                     2
humidityhighpercent             2
humidityavgpercent              2
humiditylowpercent              2
windavgmph                      2
date                            0
temphighf                       0
templowf                        0
tempavgf                        0
events                          0
dtype: int64

#### As you can see from the output, the majority of missing data is in one column called `PrecipitationSumInches`. What's the number of missing values in this column in ratio to its total number of rows?

In [23]:
# Number of missing values in [precipitationsuminches]:
null_precipitationsuminches = df.precipitationsuminches.isna().sum()
null_precipitationsuminches

124

In [24]:
# Percentage of null values in [precipitationsuminches] over total:
pct_null_precipitationsuminches = (null_precipitationsuminches * 100)/len(missing_values)
pct_null_precipitationsuminches

91.17647058823529

Almost 10% data missing! Therefore, we prefer to remove this column instead of filling its missing values.

#### Remove this column from `austin` using the `.drop()` function. Use the `inplace=True` argument.

*Hints:*

* By supplying `inplace=True` to `drop()`, the original dataframe object will be changed in place and the function will return `None`. In contrast, if you don't supply `inplace=True`, which is equivalent to supplying `inplace=False` because `False` is the default value, the original dataframe object will be kept and the function returns a copy of the transformed dataframe object. In the latter case, you'll have to assign the returned object back to your variable.

* Also, since you are dropping a column instead of a row, you'll need to supply `axis=1` to `drop()`.

[Reference for `pandas.DataFrame.drop`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.drop.html)

In [25]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1319 entries, 0 to 1318
Data columns (total 21 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   date                        1319 non-null   object 
 1   temphighf                   1319 non-null   int64  
 2   tempavgf                    1319 non-null   int64  
 3   templowf                    1319 non-null   int64  
 4   dewpointhighf               1312 non-null   float64
 5   dewpointavgf                1312 non-null   float64
 6   dewpointlowf                1312 non-null   float64
 7   humidityhighpercent         1317 non-null   float64
 8   humidityavgpercent          1317 non-null   float64
 9   humiditylowpercent          1317 non-null   float64
 10  sealevelpressurehighinches  1316 non-null   float64
 11  sealevelpressureavginches   1316 non-null   float64
 12  sealevelpressurelowinches   1316 non-null   float64
 13  visibilityhighmiles         1307 

In [26]:
# Removing (dropping) [precipitationsuminches] column for high pct of null values:
df.drop("precipitationsuminches", axis=1, inplace=True)

In [27]:
# We see no [precipitationsuminches]
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1319 entries, 0 to 1318
Data columns (total 20 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   date                        1319 non-null   object 
 1   temphighf                   1319 non-null   int64  
 2   tempavgf                    1319 non-null   int64  
 3   templowf                    1319 non-null   int64  
 4   dewpointhighf               1312 non-null   float64
 5   dewpointavgf                1312 non-null   float64
 6   dewpointlowf                1312 non-null   float64
 7   humidityhighpercent         1317 non-null   float64
 8   humidityavgpercent          1317 non-null   float64
 9   humiditylowpercent          1317 non-null   float64
 10  sealevelpressurehighinches  1316 non-null   float64
 11  sealevelpressureavginches   1316 non-null   float64
 12  sealevelpressurelowinches   1316 non-null   float64
 13  visibilityhighmiles         1307 

#### Next we will perform linear interpolation of the missing data.

This means that we will use a linear algorithm to estimate the missing data. Linear interpolation assumes that there is a straight line between the points and the missing point will fall on that line. This is a good enough approximation for weather related data. Weather related data is typically a time series. Therefore, we do not want to drop rows from our data if possible. It is prefereable to estimate the missing values rather than remove the rows. However, if you have data from a single point in time, perhaps a better solution would be to remove the rows.

If you would like to read more about linear interpolation, you can do so [here](https://en.wikipedia.org/wiki/Linear_interpolation).

In the following cell, use the `.interpolate()` function on the entire dataframe. This time pass the `inplace=False` argument to the function and assign the interpolated dataframe to a new variable called `austin_fixed` so that we can compare with `austin`.

In [28]:
# Your code here
df_fixed = df.interpolate(inplace=False)

#### Check to make sure `austin_fixed` contains no missing data. Also check `austin` - it still contains missing data.

In [29]:
# Checking df_fixed to see if it has nulls
df_fixed.isna().sum()

date                          0
temphighf                     0
tempavgf                      0
templowf                      0
dewpointhighf                 0
dewpointavgf                  0
dewpointlowf                  0
humidityhighpercent           0
humidityavgpercent            0
humiditylowpercent            0
sealevelpressurehighinches    0
sealevelpressureavginches     0
sealevelpressurelowinches     0
visibilityhighmiles           0
visibilityavgmiles            0
visibilitylowmiles            0
windhighmph                   0
windavgmph                    0
windgustmph                   0
events                        0
dtype: int64

# Challenge 3 - Processing the `Events` Column

#### Our dataframe contains one true text column - the Events column. We should evaluate this column to determine how to process it.

Use the `value_counts()` function to evaluate the contents of this column

In [45]:
df.shape

(1319, 20)

In [30]:
# [Events] value counts:
df.events.value_counts()

events
                             903
Rain                         192
Rain , Thunderstorm          137
Fog , Rain , Thunderstorm     33
Fog                           21
Thunderstorm                  17
Fog , Rain                    14
Rain , Snow                    1
Fog , Thunderstorm             1
Name: count, dtype: int64

Reading the values of `Events` and reflecting what those values mean in the context of data, you realize this column indicates what weather events had happened in a particular day.

#### What is the largest number of events happened in a single day? Enter your answer in the next cell.

In [31]:
# Your answer:
# Rain is the most frequent event

#### We want to transform the string-type `Events` values to the numbers. This will allow us to apply machine learning algorithms easily.

How? We will create a new column for each type of events (i.e. *Rain*, *Snow*, *Fog*, *Thunderstorm*. In each column, we use `1` to indicate if the corresponding event happened in that day and use `0` otherwise.

Below we provide you a list of all event types. Loop the list and create a dummy column with `0` values for each event in `austin_fixed`. To create a new dummy column with `0` values, simply use `austin_fixed[event] = 0`.

In [32]:
event_list = ['Snow', 'Fog', 'Rain', 'Thunderstorm']

# Your code here
for event in event_list:
  df_fixed['event'] = 0

# Print your new dataframe to check whether new columns have been created:

df_fixed.head()

date  temphighf  tempavgf  templowf  dewpointhighf  dewpointavgf  \
0  2013-12-21         74        60        45           67.0          49.0   
1  2013-12-22         56        48        39           43.0          36.0   
2  2013-12-23         58        45        32           31.0          27.0   
3  2013-12-24         61        46        31           36.0          28.0   
4  2013-12-25         58        50        41           44.0          40.0   

   dewpointlowf  humidityhighpercent  humidityavgpercent  humiditylowpercent  \
0          43.0                 93.0                75.0                57.0   
1          28.0                 93.0                68.0                43.0   
2          23.0                 76.0                52.0                27.0   
3          21.0                 89.0                56.0                22.0   
4          36.0                 86.0                71.0                56.0   

   sealevelpressurehighinches  sealevelpressureavginches  \
0                       29.86                      29.68   
1                       30.41                      30.13   
2                       30.56                      30.49   
3                       30.56                      30.45   
4                       30.41                      30.33   

   sealevelpressurelowinches  visibilityhighmiles  visibilityavgmiles  \
0                      29.59                 10.0                 7.0   
1                      29.87                 10.0                10.0   
2                      30.41                 10.0                10.0   
3                      30.30                 10.0                10.0   
4                      30.27                 10.0                10.0   

   visibilitylowmiles  windhighmph  windavgmph  windgustmph  \
0                 2.0         20.0         4.0         31.0   
1                 5.0         16.0         6.0         25.0   
2                10.0          8.0         3.0         12.0   
3                 7.0         12.0         4.0         20.0   
4                 7.0         10.0         2.0         16.0   

                events  event  
0  Rain , Thunderstorm      0  
1                           0  
2                           0  
3                           0  
4                           0

In [33]:
df_fixed.event.unique()

array([0])

#### Next, populate the actual values in the dummy columns of  `austin_fixed`.

You will check the *Events* column. If its string value contains `Rain`, then the *Rain* column should be `1`. The same for `Snow`, `Fog`, and `Thunderstorm`.

*Hints:*

* Use [`pandas.Series.str.contains()`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.contains.html) to create the value series of each new column.

* What if the values you populated are booleans instead of numbers? You can cast the boolean values to numbers by using `.astype(int)`. For instance, `pd.Series([True, True, False]).astype(int)` will return a new series with values of `[1, 1, 0]`.

In [34]:
# Populate the actual values in the dummy columns based on the Events column
for event in event_list:
    df_fixed[event] = df_fixed['events'].str.contains(event).astype(int)

In [35]:
df_fixed.event.unique()

array([0])

#### Print out `austin_fixed` to check if the event columns are populated with the intended values

In [46]:
# Your code here
df_fixed.head()

date  temphighf  tempavgf  templowf  dewpointhighf  dewpointavgf  \
0  735223         74        60        45           67.0          49.0   
1  735224         56        48        39           43.0          36.0   
2  735225         58        45        32           31.0          27.0   
3  735226         61        46        31           36.0          28.0   
4  735227         58        50        41           44.0          40.0   

   dewpointlowf  humidityhighpercent  humidityavgpercent  humiditylowpercent  \
0          43.0                 93.0                75.0                57.0   
1          28.0                 93.0                68.0                43.0   
2          23.0                 76.0                52.0                27.0   
3          21.0                 89.0                56.0                22.0   
4          36.0                 86.0                71.0                56.0   

   sealevelpressurehighinches  sealevelpressureavginches  \
0                       29.86                      29.68   
1                       30.41                      30.13   
2                       30.56                      30.49   
3                       30.56                      30.45   
4                       30.41                      30.33   

   sealevelpressurelowinches  visibilityhighmiles  visibilityavgmiles  \
0                      29.59                 10.0                 7.0   
1                      29.87                 10.0                10.0   
2                      30.41                 10.0                10.0   
3                      30.30                 10.0                10.0   
4                      30.27                 10.0                10.0   

   visibilitylowmiles  windhighmph  windavgmph  windgustmph  \
0                 2.0         20.0         4.0         31.0   
1                 5.0         16.0         6.0         25.0   
2                10.0          8.0         3.0         12.0   
3                 7.0         12.0         4.0         20.0   
4                 7.0         10.0         2.0         16.0   

                events  event  Snow  Fog  Rain  Thunderstorm  
0  Rain , Thunderstorm      0     0    0     1             1  
1                           0     0    0     0             0  
2                           0     0    0     0             0  
3                           0     0    0     0             0  
4                           0     0    0     0             0

#### If your code worked correctly, now we can drop the `Events` column as we don't need it any more.

In [48]:
# Your code here
df_fixed.drop(columns="events")

date  temphighf  tempavgf  templowf  dewpointhighf  dewpointavgf  \
0     735223         74        60        45           67.0          49.0   
1     735224         56        48        39           43.0          36.0   
2     735225         58        45        32           31.0          27.0   
3     735226         61        46        31           36.0          28.0   
4     735227         58        50        41           44.0          40.0   
...      ...        ...       ...       ...            ...           ...   
1314  736537        103        89        75           71.0          67.0   
1315  736538        105        91        76           71.0          64.0   
1316  736539        107        92        77           72.0          64.0   
1317  736540        106        93        79           70.0          68.0   
1318  736541         99        88        77           66.0          61.0   

      dewpointlowf  humidityhighpercent  humidityavgpercent  \
0             43.0                 93.0                75.0   
1             28.0                 93.0                68.0   
2             23.0                 76.0                52.0   
3             21.0                 89.0                56.0   
4             36.0                 86.0                71.0   
...            ...                  ...                 ...   
1314          61.0                 82.0                54.0   
1315          55.0                 87.0                54.0   
1316          55.0                 82.0                51.0   
1317          63.0                 69.0                48.0   
1318          54.0                 64.0                43.0   

      humiditylowpercent  sealevelpressurehighinches  \
0                   57.0                       29.86   
1                   43.0                       30.41   
2                   27.0                       30.56   
3                   22.0                       30.56   
4                   56.0                       30.41   
...                  ...                         ...   
1314                25.0                       30.04   
1315                20.0                       29.97   
1316                19.0                       29.91   
1317                27.0                       29.96   
1318                22.0                       30.04   

      sealevelpressureavginches  sealevelpressurelowinches  \
0                         29.68                      29.59   
1                         30.13                      29.87   
2                         30.49                      30.41   
3                         30.45                      30.30   
4                         30.33                      30.27   
...                         ...                        ...   
1314                      29.97                      29.88   
1315                      29.90                      29.81   
1316                      29.86                      29.79   
1317                      29.91                      29.87   
1318                      29.97                      29.91   

      visibilityhighmiles  visibilityavgmiles  visibilitylowmiles  \
0                    10.0                 7.0                 2.0   
1                    10.0                10.0                 5.0   
2                    10.0                10.0                10.0   
3                    10.0                10.0                 7.0   
4                    10.0                10.0                 7.0   
...                   ...                 ...                 ...   
1314                 10.0                10.0                10.0   
1315                 10.0                10.0                10.0   
1316                 10.0                10.0                10.0   
1317                 10.0                10.0                10.0   
1318                 10.0                10.0                10.0   

      windhighmph  windavgmph  windgustmph  event  Snow  Fog  Rain  \
0            20.0    

# Challenge 4 - Processing The `Date` Column

The `Date` column is another non-numeric field in our dataset. A value in that field looks like `'2014-01-06'` which consists of the year, month, and day connected with hyphens. One way to convert the date string to numerical is using a similar approach as we used for `Events`, namely splitting the column into numerical `Year`, `Month`, and `Day` columns. In this challenge we'll show you another way which is to use the Python `datetime` library's `toordinal()` function. Depending on what actual machine learning analysis you will conduct, each approach has its pros and cons. Our goal today is to practice data preparation so we'll skip the discussion here.

Here you can find the [reference](https://docs.python.org/3/library/datetime.html) and [example](https://stackoverflow.com/questions/39846918/convert-date-to-ordinal-python) for `toordinal`. The basic process is to first convert the string to a `datetime` object using `datetime.datetime.strptime`, then convert the `datetime` object to numerical using `toordinal`.

#### In the cell below, convert the `Date` column values from string to numeric values using `toordinal()`.

In [39]:
# Your code here
df_fixed['date'] = pd.to_datetime(df_fixed['date']).apply(pd.Timestamp.toordinal)

#### Print `austin_fixed` to check your `Date` column.

In [40]:
df_fixed.head(5)

date  temphighf  tempavgf  templowf  dewpointhighf  dewpointavgf  \
0  735223         74        60        45           67.0          49.0   
1  735224         56        48        39           43.0          36.0   
2  735225         58        45        32           31.0          27.0   
3  735226         61        46        31           36.0          28.0   
4  735227         58        50        41           44.0          40.0   

   dewpointlowf  humidityhighpercent  humidityavgpercent  humiditylowpercent  \
0          43.0                 93.0                75.0                57.0   
1          28.0                 93.0                68.0                43.0   
2          23.0                 76.0                52.0                27.0   
3          21.0                 89.0                56.0                22.0   
4          36.0                 86.0                71.0                56.0   

   sealevelpressurehighinches  sealevelpressureavginches  \
0                       29.86                      29.68   
1                       30.41                      30.13   
2                       30.56                      30.49   
3                       30.56                      30.45   
4                       30.41                      30.33   

   sealevelpressurelowinches  visibilityhighmiles  visibilityavgmiles  \
0                      29.59                 10.0                 7.0   
1                      29.87                 10.0                10.0   
2                      30.41                 10.0                10.0   
3                      30.30                 10.0                10.0   
4                      30.27                 10.0                10.0   

   visibilitylowmiles  windhighmph  windavgmph  windgustmph  \
0                 2.0         20.0         4.0         31.0   
1                 5.0         16.0         6.0         25.0   
2                10.0          8.0         3.0         12.0   
3                 7.0         12.0         4.0         20.0   
4                 7.0         10.0         2.0         16.0   

                events  event  Snow  Fog  Rain  Thunderstorm  
0  Rain , Thunderstorm      0     0    0     1             1  
1                           0     0    0     0             0  
2                           0     0    0     0             0  
3                           0     0    0     0             0  
4                           0     0    0     0             0

# Challenge 5 - Sampling and Holdout Sets

#### Now that we have processed the data for machine learning, we will separate the data to test and training sets.

We first train the model using only the training set. We check our metrics on the training set. We then apply the model to the test set and check our metrics on the test set as well. If the metrics are significantly more optimal on the training set, then we know we have overfit our model. We will need to revise our model to ensure it will be more applicable to data outside the test set.

#### In the next cells we will separate the data into a training set and a test set using the `train_test_split()` function in scikit-learn.

When using `scikit-learn` for machine learning, we first separate the data to predictor and response variables. This is the standard way of passing datasets into a model in `scikit-learn`. The `scikit-learn` will then find out whether the predictors and responses fit the model.

In the next cell, assign the `TempAvgF` column to `y` and the remaining columns to `X`. Your `X` should be a subset of `austin_fixed` containing the following columns:

```['Date',
 'TempHighF',
 'TempLowF',
 'DewPointHighF',
 'DewPointAvgF',
 'DewPointLowF',
 'HumidityHighPercent',
 'HumidityAvgPercent',
 'HumidityLowPercent',
 'SeaLevelPressureHighInches',
 'SeaLevelPressureAvgInches',
 'SeaLevelPressureLowInches',
 'VisibilityHighMiles',
 'VisibilityAvgMiles',
 'VisibilityLowMiles',
 'WindHighMPH',
 'WindAvgMPH',
 'WindGustMPH',
 'Snow',
 'Fog',
 'Rain',
 'Thunderstorm']```

 Your `y` should be a subset of `austin_fixed` containing one column `TempAvgF`.

In [41]:
y = df_fixed.tempavgf

In [42]:
X = df_fixed[['date', 'temphighf', 'templowf', 'dewpointhighf', 'dewpointavgf', 'dewpointlowf',
              'humidityhighpercent', 'humidityavgpercent', 'humiditylowpercent',
              'sealevelpressurehighinches', 'sealevelpressureavginches', 'sealevelpressurelowinches',
              'visibilityhighmiles', 'visibilityavgmiles', 'visibilitylowmiles',
              'windhighmph', 'windavgmph', 'windgustmph', 'Snow', 'Fog', 'Rain', 'Thunderstorm']]

In the next cell, import `train_test_split` from `sklearn.model_selection`

In [43]:
# Your code here:
from sklearn.model_selection import train_test_split

Now that we have split the data to predictor and response variables and imported the `train_test_split()` function, split `X` and `y` into `X_train`, `X_test`, `y_train`, and `y_test`. 80% of the data should be in the training set and 20% in the test set. `train_test_split()` reference can be accessed [here](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html).


Enter your code in the cell below:

In [44]:
#Your code here:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

#### Congratulations! Now you have finished the preparation of the dataset!